### PR TITLE
Sink Reshape below BatchNorm.

### PR DESCRIPTION
**Pull request by ugaresuyash**
[graph optimization] Sink reshape below BatchNorm 

Summary: 
Sinking reshape below BatchNorm enables merge of Conv and BatchNorm.
Sink is performed only under certain conditions:
 - Reshape does not alter number of Channels and Batches
 - ChannelIdx of BatchNorm is indexing to 'C' in NH[W]C or NCH[W]

Change-Id: I380a12203d214ca3d1c03aa8ecbb2dd840a71ce8

Documentation:

[Optional Fixes #issue]

Test Plan: Added a unittest.

Please see a detailed explanation of how to fill out the fields in the relevant sections in PULL_REQUEST.md.

### CODE REVIEW

File: lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp [hunk 1]
```diff
@@ -409,6 +409,41 @@ bool SinkCode::run(Function *F, const CompilationContext &cctx) {
   // For each node:
   for (auto &N : nodes) {
     auto *node = &N;
+    // Sink Reshape below BatchNormalization.
+    if (auto *BN = dyn_cast<BatchNormalizationNode>(node)) {
+      if (auto *RS = dyn_cast<ReshapeNode>(BN->getInput())) {
+        auto inDims = RS->getInput().dims();
+        auto outDims = RS->getType(0)->dims();
```
Comment (ayermolo) [inline lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp:416]: We already know what RS is. can just getResult and type from there.

Comment (ugaresuyash) [reply lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp:416]: @ayermolo, I am not sure I understood you point. Could you please elaborate?

**Comment by ugaresuyash** [comment]
@jfix71, Could you please review this?

File: lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp [hunk 2]
```diff
@@ -409,6 +409,41 @@ bool SinkCode::run(Function *F, const CompilationContext &cctx) {
...
+        // Reshape should not change the Channel Idx dimensions.
+        // Num input channels is equal to num output channels.
+        // BatchNorm ChannelIdx is C in NHWC or NWC format.
```
Comment (jfix71) [inline lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp:419]: We also use NCHW, so we don't know for sure it will always be the final dimension. Perhaps we could just check that dim `BN->getChannelIdx()` is the same before and after, and that there are the same number of dimensions before and after?

File: lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp [hunk 3]
```diff
@@ -409,6 +409,41 @@ bool SinkCode::run(Function *F, const CompilationContext &cctx) {
...
+        if (!RS->hasOneUse()) {
+          continue;
+        }
+
+        unsigned_t newChannelIdx = inDims.size() - 1;
+        auto *newBN = F->createBatchNormalization(
+            BN->getName(), RS->getInput(), BN->getBias(), BN->getScale(),
+            BN->getMean(), BN->getVar(), newChannelIdx, BN->getEpsilon(),
+            BN->getMomentum());
+        RS->setNthInput(0, newBN);
```
Comment (jfix71) [inline lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp:439]:
```suggestion
        RS->setNthInput(ReshapeNode::InputIdx, newBN);
```

File: lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp (hunk 2 above)
Comment (ugaresuyash) [reply lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp:419]: "check that dim BN->getChannelIdx() is the same before and after" - Makes sense.
"same number of dimensions before and after" - may not be the case always. Reshape could flatten HW dimensions to W/H, and it does not affect the optimization.

File: lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp (hunk 3 above)
Comment (ugaresuyash) [reply lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp:440]: Done.

File: lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp [hunk 4]
```diff
@@ -409,6 +409,41 @@ bool SinkCode::run(Function *F, const CompilationContext &cctx) {
   // For each node:
   for (auto &N : nodes) {
     auto *node = &N;
+    // Sink Reshape below BatchNormalization.
+    if (auto *BN = dyn_cast<BatchNormalizationNode>(node)) {
```
[the remark this answers is not in the document]
Comment (ugaresuyash) [reply lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp:414]: Makes sense, done.

File: lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp (hunk 2 above)
Comment (jfix71) [reply lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp:423]: Ok sure, but we still should be able to modify this to handle NCHW too, right? We can sink the reshape if C stays the same. So for NHWC/NWC we can always check the final dimension like you are. For NCHW/NCW we can always check the second dimension.

Comment (ugaresuyash) [reply lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp]: Ahh... understood. Done.

Comment (ugaresuyash) [reply lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp]: @jfix71 , Were you able to take a look at it?

File: lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp [hunk 5]
```diff
@@ -410,31 +410,72 @@ bool SinkCode::run(Function *F, const CompilationContext &cctx) {
...
+        auto outDims = RS->getResult().dims();
+
+        // Reshape should not change the BatchNorm ChannelIdx dimensions.
+
+        // Check for NH[W]C layout.
+        if (outDims[BN->getChannelIdx()] == outDims.size() - 1) {
+          if (inDims.size() - 1 != outDims.size() - 1) {
+            continue;
+          }
+        }
```
Comment (jfix71) [inline lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp:428]: I'm a little confused -- shouldn't this be something like 
```suggestion
        if (BN->getChannelIdx() == outDims.size() - 1 && 
            inDims[size() - 1] == outDims[size() -1]) {
          continue;
        }
```

File: lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp [hunk 6]
```diff
@@ -410,31 +410,72 @@ bool SinkCode::run(Function *F, const CompilationContext &cctx) {
...
+
+        // Check for NCH[W] layout.
+        if (outDims[BN->getChannelIdx()] == outDims[1]) {
+          if (inDims[1] != outDims[1]) {
+            continue;
+          }
+        }
```
Comment (jfix71) [inline lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp:435]: Same -- shouldn't this be something like
```suggestion
        // Note: 1 here is for C in NCH[W]
        if (BN->getChannelIdx() != 1 && inDims[1] == outDims[1]) {
          continue;
        }
```

File: tests/unittests/GraphOptzTest.cpp [hunk 7]
```diff
@@ -184,6 +184,34 @@ TEST_F(GraphOptz, liveCodeNotEliminated) {
   EXPECT_EQ(mod_.getPlaceholders().size(), 3);
 }
 
+TEST_F(GraphOptz, optimizeBatchNormAfterConvAndReshape) {
```
Comment (jfix71) [inline tests/unittests/GraphOptzTest.cpp:187]: Test LG, can we add one for NCHW, and maybe also one where we think the opt shouldn't kick in? I think the conditions you have right now aren't right, and so it would actually cause the opt to kick in when it shouldn't and should fail a test. Like make the batch, channel, H all the same value to make sure we're comparing the right indexing.

File: lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp (hunk 5 above)
Comment (ugaresuyash) [reply lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp:428]: @jfix71, Thanks for pointing put the mistake. I have fixed this for NH[W]C layout.

File: lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp (hunk 6 above)
Comment (ugaresuyash) [reply lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp:435]: Assuming Conv -> Reshape -> BatchNorm sequence, for NCH[W], if we allow sink of Reshape below BatchNorm then, BatchNorm will fold into Conv which should not happen. BatchNorm folding into Convolution is allowed only when the layout is NHWC. For now I am not allowing the optimization to kick in for NCHW case.

File: tests/unittests/GraphOptzTest.cpp (hunk 7 above)
Comment (ugaresuyash) [reply tests/unittests/GraphOptzTest.cpp:187]: I have added 3 tests: two positive tests and one negative test.

File: lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp (hunk 6 above)
Comment (jfix71) [reply lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp]: >BatchNorm folding into Convolution is allowed only when the layout is NHWC.

@ugaresuyash Maybe I'm missing something, but why should that be the case? NHWC vs NCHW is just about the layout of the data in memory, why would the fusion of BN require NHWC?

Assuming it is the case for some reason, it seems a bit scary to rely on there being a reshape there to prevent the conv-batch norm fusion from kicking in. Seems to me like we should allow this reshape sinking to be done regardless of the layout, but then also update the conv-batch norm fusion to only kick in when we have an NHWC.

Comment (ugaresuyash) [reply lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp]: @jfix, BatchNorm fold assumes that the layout has to be NHWC when fusing BatchNorm parameters into Convolution weights.  Please refer to 'normalizeWeights()'.

What you say also makes sense. I will look into adding the sink optimization and then prevent Conv + BatchNorm fusion.

Comment (jfix71) [reply lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp]: Oh, you just meant what glow currently supports, got it.

Comment (ugaresuyash) [reply lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp]: @jfix71 ,
I did attempt to allow sink of reshape below batchnorm for NCHW case. Sink works well. However, during normalizeWeights(), I see the channel idx of BN as 3 instead of 1 which was set in the test body. Something strange is happening. Either there is some other issue, potentially, which is not preserving the BN channel idx. Or may be I am missing something.

For now, for my internal work, I think NHWC serves the purpose. Can we land NHWC for now and then later create an issue to handle the NCHW case separately?

Comment (ugaresuyash) [reply lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp]: @jfix71, I have made the necessary changes for adding NCHW case and a corresponding test. 
Please take look at it.
